### ISSUE TITLE
Passing mat4 from vertex shader to fragment

### ISSUE THREAD

**Issue by ryhor-spivak** [gfx-rs/wgpu-native#143]
It's causing panics but only on Metal backend:

thread '<unnamed>' panicked at 'Internal { stage: VERTEX, error: "Metal: Compilation failed: \n\nprogram_source:36:43: error: type 'metal::float4x4' (aka 'matrix<float, 4, 4>') is not valid for attribute 'flat'\n  metal::float4x4 member1 [[user(loc0), flat]];\n                     ^~~~\nprogram_source:38:8: error: invalid return type 'main1Output' for vertex function\nvertex main1Output main1(\n    ^\n" }', src/device.rs:566:9

Shader is submitted in SPIR-V, compiled from this GLSL:

    #version 450

    layout( location = 0 ) in vec3 attrib_pos;

    out gl_PerVertex
    {
        vec4 gl_Position;
    };

    layout( location = 0 ) out mat4 var_m;

    void main()
    {
        gl_Position = vec4(attrib_pos,1.);
        var_m = mat4(
            1., 0., 0., 0.,
            0., 1., 0., 0.,
            0., 0., 1., 0.,
            0., 0., 0., 1.
            );
    }

Is there any limitations on matrix varyings with Metal? With Vulkan same shader works as expected.

**Comment by ryhor-spivak** [comment]
This is reasonable. General version of matrix inverse is an overkill. In our case it is used only because of laziness as matrices are translation-rotation-scale which can be inverted with a just few operations.

Just hit another bump with SPIR-V shaders on current version: OpExtInst Radians compiled from glsl radians() call. Loading such shader results in a panic with a very strange error for such case: ExpressionAlreadyInScope.

**Comment by kvark** [comment]
right, that looks like https://github.com/gfx-rs/naga/pull/827, which unfortunately hasn't landed.
cc @expenses
